Interprétation LaTeX par le bot lui-même

Niilyx opened the pull request:
Pour ça j'utilise [MathJax](https://www.npmjs.com/package/mathjax) pour le parsing des formules LaTeX et [Sharp](https://www.npmjs.com/package/sharp) pour la conversion en un .png

Il reste un problème à fix, sur certaines formules test, MathJax échoue le parsing et en réessayant tout de suite après, il y parvient.

Exemple de formule problématique :
```
\begin{align*}
\sum\limits_{k=1}^n u_k
&= \sum\limits_{k=1}^n (\frac{2}{k} - \frac{2}{k+2}) \\ 
&= \sum\limits_{k=1}^n \frac{2}{k} - \sum\limits_{k=1}^n \frac{2}{k+2} \\
&= \sum\limits_{k=1}^n \frac{2}{k} - \sum\limits_{k=3}^{n+2} \frac{2}{k} \\
&= \sum\limits_{k=3}^n \frac{2}{k} +2 +1 - (\sum\limits_{k=3}^{n} \frac{2}{k} + \frac{2}{n+1} + \frac{2}{n+2}) \\
&= \textcolor{blue}{\sum\limits_{k=3}^n \frac{2}{k}} + 3 \textcolor{blue}{- \sum\limits_{k=3}^{n} \frac{2}{k}} - \frac{2}{n+1} - \frac{2}{n+2} \\
&= 3 - \frac{2}{n+1} - \frac{2}{n+2}     \end{align*}
```
c'est-à-dire : 
![image](https://github.com/horizon-efrei/HorizonBot/assets/43385037/223e6a94-1f2d-41b3-a850-6d0757f0344f)







Si vous avez une idée pour fix ça, où une meilleure librairie de parsing à proposer, je suis tout ouïe

Niilyx committed:
test mathjax

Niilyx committed:
+ MathJax
+ Sharp

Manque plus qu'à fix un problème

Niilyx committed:
Opti imports

Niilyx committed:
Commentaires + petit message d'erreur

Niilyx committed:
Error handling de la formule elle-même + retry automatique du parsing

Niilyx committed:
+ Support de packages basiques de TeX (ams, mathtools...)
+ Réductions de l'image des très petites formules (moins de 40 caractères)

Niilyx committed:
Restoration main.ts

Niilyx commented:
Bientôt y aura quelque chose d'autre qui arrivera :eyes:

Niilyx committed:
+TeX packages +Extend max characters for width change

Niilyx committed:
ESLint

noftaly commented on src/commands/General/latex.ts:
```diff
@@ -28,8 +25,45 @@ const equationModal = new ModalBuilder()
+  if (node.children) {
+    for (const child of node.children) {
+      if (child.kind === 'defs')
+continue;
```
`eslint/indent` est buggé, mais ici faut bien indenter le `continue`.

noftaly commented on src/commands/General/latex.ts:
```diff
@@ -28,8 +25,45 @@ const equationModal = new ModalBuilder()
+    },
+    tex: { packages: { '[+]': ['ams', 'color', 'mathtools', 'physics', 'unicode', 'textmacros'] } },
+  }).then((readyMJ: object) => {
+    // Explicitement bypass le readonly et effectivement faire un await dans à l'initialisation
```
[Commentaire en anglais please](https://github.com/horizon-efrei/HorizonBot/blob/bcda527400e960b4b3e91eaaf8a3eb0c15793ca2/CONTRIBUTING.md?plain=1#L32-L34)

noftaly commented on src/commands/General/latex.ts:
```diff
@@ -63,6 +98,63 @@ export class LatexCommand extends HorizonCommand<typeof config> {
+    if (svgError)
+      throw new EvalError(svgError);
+
+    const pad = equation.length > 40 ? 45 : 4;
+    return sharp(Buffer.from(svgText))
+      .resize({
+        width: equation.length > 40 ? 1200 : Math.max(Math.floor(8 * equation.length), 64),
+      }) // Sinon l'image est trop petite et pixellisée
+      .extend({
+        top: pad,
+        bottom: pad,
+        left: pad,
+        right: pad,
```
J'ai un peu expérimenté de mon côté, et tu penses quoi ce ça ? C'est un bon compromis je pense, qui scale en fonction de la taille de l'équation (avec un max à 1024). Ça rend pas trop mal, je te laisse tester

```suggestion
    return sharp(Buffer.from(svgText))
      .resize({ width: Math.min(Math.floor(6 * equation.length), 1024) })
      .extend({
        top: 10,
        bottom: 10,
        left: 10,
        right: 10,
```

noftaly commented on src/commands/General/latex.ts:
```diff
@@ -63,6 +98,63 @@ export class LatexCommand extends HorizonCommand<typeof config> {
+        right: pad,
+        background: '#FFF',
+      }) // Padding
+      .flatten({ background: '#FFF' }) // Arrière-plan de couleur
```
Commentaire en anglais please

noftaly commented on src/commands/General/latex.ts:
```diff
@@ -63,6 +98,63 @@ export class LatexCommand extends HorizonCommand<typeof config> {
+      .toBuffer();
+  }
+
+  private async _tryToGenerate(replyTo: CommandInteraction<'cached'> | ModalSubmitInteraction<'cached'>, equation: string, attempt: number): Promise<void> {
+    if (attempt > 3)
+      throw new Error('MathJax retry limit reached');
+
+    try {
+      const result = await this._parseAndGenerate(equation);
+      await replyTo.reply({ files: [new AttachmentBuilder(result).setName('output.png')] });
+    } catch (e) {
+      if (e.toString().includes('MathJax retry'))
+        await this._tryToGenerate(replyTo, equation, attempt + 1);
+      else
+        throw e;
+    }
   }
```
L'idéal serait que cette fonction return un `Promise<Buffer>` et que ce soit le `chatInputRun` qui s'occupe de la lancer.
Ça a plusieurs avantages, déjà pas besoin de passer le replyTo à cette fonction, et ensuite si on veut la réutiliser pour qqch d'autre ce sera plus simple de la réutiliser/l'extraire dans un `utils/`.
Tu peux aussi en profiter pour rajouter une valeur par défaut au paramètre attempt : `attempt = 0`, comme ca pas besoin de la spécifier quand on l'appelle depuis `chatInputRun`.

noftaly commented on src/commands/General/latex.ts:
```diff
@@ -63,6 +98,63 @@ export class LatexCommand extends HorizonCommand<typeof config> {
+      await this._tryToGenerate(replyTo, equation, 0);
+    } catch (e) {
+      if (e instanceof EvalError) {
+        await replyTo.reply({
+          content: "Ça n'a pas l'air d'être une équation valide ! Le message d'erreur LaTeX est :\n"
+            + `\`${e.message}\``,
+          ephemeral: true,
+        });
+      } else {
+        await replyTo.reply({
+          content: "L'image de la formule n'a pas pu être générée.",
+          ephemeral: true,
+        });
```
Il faudrait les passer dans la configuration.
Le deuxième est straightforward, pour le premier faudra utiliser `pupa`. je te laisse regarder comment est-ce qu'on fait dans d'autres commandes, mais il faudrait en gros rajouter deux entrées dans [l'objet de config](https://github.com/horizon-efrei/HorizonBot/blob/bcda527400e960b4b3e91eaaf8a3eb0c15793ca2/config/commands/general.ts#L71-L72), puis ici faire un `this.messages.{le nom de l'entrée}` (l'autocomplétion TypeScript pourra bien t'aider).

Exemple dans le fichier de la commande ping: [commande](https://github.com/horizon-efrei/HorizonBot/blob/bcda527400e960b4b3e91eaaf8a3eb0c15793ca2/src/commands/General/ping.ts#L26) - [config](https://github.com/horizon-efrei/HorizonBot/blob/bcda527400e960b4b3e91eaaf8a3eb0c15793ca2/config/commands/general.ts#L98C62-L98C62)

noftaly commented on src/commands/General/latex.ts:
```diff
@@ -28,8 +25,45 @@ const equationModal = new ModalBuilder()
+
 @ApplyOptions<HorizonCommand.Options>(config)
 export class LatexCommand extends HorizonCommand<typeof config> {
+  private readonly _mathJax: any = require('mathjax').init({
```
Utilise un `import` et pas un `require`
Tu peux notamment faire un `import { init as initMathJax } from 'mathjax';` par exemple, puis ici appeler la fonction  `initMathjax`. Hésite pas à rajouter des disable d'eslint ou de typescript car mathjax est typé avec le cul

noftaly reviewed (changes requested):
Merci beaucoup pour la PR et d'avoir pris le temps de trifouiller MathJax !
Quelques petites suggestions à prendre en compte, hésite pas à me contacter sur discord si t'as des questions, ou à répondre aux commentaires direct sur GitHub.

Btw le lint ne fonctionne pas, t'as du lancer `npm run lint:fix` mais ca ne suffit pas car il y a des choses qu'il ne peut pas régler tout seul. Je te laisse voir avec `npm run lint` ou en regardant [l'action GitHub qui a fail](https://github.com/horizon-efrei/HorizonBot/actions/runs/6818311391/job/18573616746?pr=97#step:6:11)

Niilyx committed:
Update src/commands/General/latex.ts

Co-authored-by: Elliot <elliot.maisl@gmail.com>

Niilyx committed:
+ npm audit fix
+ import instead of require MathJax
+ eslint ignores and ts-expect-error
+ Corrected indentation
+ Made function _tryToGenerate more independent
+ English comments :trollface:

Niilyx committed:
+ Template literals for error messages in /latex implemented

noftaly reviewed (approved):
LGTM 🚀
Merci beaucoup !

Niilyx committed:
Final commit:
+ No more latex API link
+ Added Thanks for Idir :)
